azurelinuxagent/common/utils/archive.py

```diff
@@ -0,0 +1,213 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# Requires Python 2.6+ and Openssl 1.0+
+#
+import os
+import re
+import shutil
+import zipfile
+
+from datetime import datetime
+
+from azurelinuxagent.common.utils import fileutil
+
+
+"""
+archive.py
+
+The module supports the archiving of guest agent state. Guest
+agent state is flushed whenever there is a incarnation change.
+The flush is archived periodically (once an hour).
+
+The process works as follows whenever a new incarnation arrives.
+
+ 1. Flush - move all state files to a new directory under
+ .../history/timestamp/.
+ 2. Archive - enumerate all directories under .../history/timestamp
+ and create a .zip file named timestamp.zip.  Delete the archive
+ directory
+ 3. Purge - glob the list .zip files, sort by timestamp in descending
+ order, keep the first 50 results, and delete the rest.
+
+... is the directory where the agent's state resides, by default this
+is /var/lib/waagent.
+
+The timestamp is an ISO8601 formatted value.
+"""
+
+MAX_ARCHIVED_STATES = 50
+
+CACHE_PATTERNS = [
+    re.compile("^(.*)\.(\d+)\.(agentsManifest)$", re.IGNORECASE),
+    re.compile("^(.*)\.(\d+)\.(manifest\.xml)$", re.IGNORECASE),
+    re.compile("^(.*)\.(\d+)\.(xml)$", re.IGNORECASE)
+]
+
+# 2018-04-06T08:21:37.142697
+# 2018-04-06T08:21:37.142697.zip
+ARCHIVE_PATTERNS_DIRECTORY = re.compile('^\d{4}\-\d{2}\-\d{2}T\d{2}:\d{2}:\d{2}\.\d+$')
+ARCHIVE_PATTERNS_ZIP       = re.compile('^\d{4}\-\d{2}\-\d{2}T\d{2}:\d{2}:\d{2}\.\d+\.zip$')
+
+
+class StateFlusher(object):
+    def __init__(self, lib_dir):
+        self._timestamp = datetime.now()
+        self._source = lib_dir
+        self._target = os.path.join(self._source, 'history', self._timestamp.isoformat())
+
+    def flush(self):
+        files = self._get_files_to_archive()
+        if len(files) == 0:
+            return
+
+        if self._mkdir():
+            self._archive(files)
+        else:
+            self._purge(files)
+
+    def _get_files_to_archive(self):
+        files = []
+        for f in os.listdir(self._source):
+            full_path = os.path.join(self._source, f)
+            for pattern in CACHE_PATTERNS:
+                m = pattern.match(f)
+                if m is not None:
+                    files.append(full_path)
+                    break
+
+        return files
+
+    def _archive(self, files):
+        for file in files:
+            fileutil.move_file(file, to_dir=self._target)
+
+    def _purge(self, files):
+        for file in files:
+            os.remove(file)
+
+    def _mkdir(self):
+        try:
+            fileutil.mkdir(self._target, mode=0o700)
+            return True
+        except:
+            return False
+
+
+# @total_ordering first appeared in Python 2.7 and 3.2
+# If there are more use cases for @total_ordering, I will
+# consider re-implementing it.
```

Comment:
TODO: Use total_ordering once RHEL/CentOS 6 and SLES 11 are EOL.

Reply:
The sooner we can drop support for 2.6, the happier we will all be. For example, any tests that expect exceptions need to be re-examined to see if self.assertRaises() should be used as a context manager.

Reply:
I'll bake the cake.  It ain't gonna happen any time soon.